File: ClassEval_T/java/solutuon/URLHandler.java
package org.example;

import java.util.HashMap;
import java.util.Map;

public class URLHandler {
    private String url;

    public URLHandler(String url) {
        this.url = url;
    }

    public String getScheme() {
        int schemeEnd = url.indexOf("://");
        if (schemeEnd != -1) {
            return url.substring(0, schemeEnd);
        }
        return null;
    }

    public String getHost() {
        int schemeEnd = url.indexOf("://");
        if (schemeEnd != -1) {
            String urlWithoutScheme = url.substring(schemeEnd + 3);
            int hostEnd = urlWithoutScheme.indexOf("/");
            if (hostEnd != -1) {
                return urlWithoutScheme.substring(0, hostEnd);
            }
            return urlWithoutScheme;
        }
        return null;
    }

    public String getPath() {
        int schemeEnd = url.indexOf("://");
        if (schemeEnd != -1) {
            String urlWithoutScheme = url.substring(schemeEnd + 3);
            int hostEnd = urlWithoutScheme.indexOf("/");
            if (hostEnd != -1) {
                return urlWithoutScheme.substring(hostEnd);
            }
        }
        return null;
    }

    public Map<String, String> getQueryParams() {
        int queryStart = url.indexOf("?");
        int fragmentStart = url.indexOf("#");
        if (queryStart != -1) {
            String queryString = url.substring(queryStart + 1, fragmentStart != -1 ? fragmentStart : url.length());
            Map<String, String> params = new HashMap<>();
            if (!queryString.isEmpty()) {
                String[] paramPairs = queryString.split("&");
                for (String pair : paramPairs) {
                    String[] keyValue = pair.split("=");
                    if (keyValue.length == 2) {
                        params.put(keyValue[0], keyValue[1]);
                    }
                }
            }
            return params;
        }
        return null;
    }

    public String getFragment() {
        int fragmentStart = url.indexOf("#");
        if (fragmentStart != -1) {
            return url.substring(fragmentStart + 1);
        }
        return null;
    }
}

File: ClassEval_T/cpp/solution/URLHandler.cpp
#include <string>
#include <map>
#include <algorithm>

class URLHandler {
public:
    explicit URLHandler(const std::string& url);

    std::string get_scheme() const;
    std::string get_host() const;
    std::string get_path() const;
    std::map<std::string, std::string> get_query_params() const;
    std::string get_fragment() const;

private:
    std::string url;
};

URLHandler::URLHandler(const std::string& url) : url(url) {}

std::string URLHandler::get_scheme() const {
    size_t scheme_end = url.find("://");
    if (scheme_end != std::string::npos) {
        return url.substr(0, scheme_end);
    }
    return "";
}

std::string URLHandler::get_host() const {
    size_t scheme_end = url.find("://");
    if (scheme_end != std::string::npos) {
        std::string url_without_scheme = url.substr(scheme_end + 3);
        size_t host_end = url_without_scheme.find("/");
        if (host_end != std::string::npos) {
            return url_without_scheme.substr(0, host_end);
        }
        return url_without_scheme;
    }
    return "";
}

std::string URLHandler::get_path() const {
    size_t scheme_end = url.find("://");
    if (scheme_end != std::string::npos) {
        std::string url_without_scheme = url.substr(scheme_end + 3);
        size_t host_end = url_without_scheme.find("/");
        if (host_end != std::string::npos) {
            return url_without_scheme.substr(host_end);
        }
    }
    return "";
}

std::map<std::string, std::string> URLHandler::get_query_params() const {
    std::map<std::string, std::string> params;
    size_t query_start = url.find("?");
    size_t fragment_start = url.find("#");
    if (query_start != std::string::npos) {
        std::string query_string = url.substr(query_start + 1, fragment_start - query_start - 1);
        if (!query_string.empty()) {
            size_t pos = 0;
            std::string token;
            while ((pos = query_string.find("&")) != std::string::npos) {
                token = query_string.substr(0, pos);
                size_t equal_pos = token.find("=");
                if (equal_pos != std::string::npos) {
                    std::string key = token.substr(0, equal_pos);
                    std::string value = token.substr(equal_pos + 1);
                    params[key] = value;
                }
                query_string.erase(0, pos + 1);
            }
            size_t equal_pos = query_string.find("=");
            if (equal_pos != std::string::npos) {
                std::string key = query_string.substr(0, equal_pos);
                std::string value = query_string.substr(equal_pos + 1);
                params[key] = value;
            }
        }
    }
    return params;
}

std::string URLHandler::get_fragment() const {
    size_t fragment_start = url.find("#");
    if (fragment_start != std::string::npos) {
        return url.substr(fragment_start + 1);
    }
    return "";
}
